### DISCUSSION TITLE
iPXE stucks once it tries to wait for user input when loaded by ARM64 U-Boot's bootefi

### DISCUSSION THREAD

**Discussion by RadxaYuntian** [ipxe/ipxe#605] [Q&A] [answered]
Recently one of our user had [issue using iPXE](https://forum.radxa.com/t/rock-pi-4c-ipxe/8554) with our ARM single board computer. In short iPXE stucks inside `shell_banner()` when it tries to wait for keyboard input. I have tested U-Boot 2022.01 with thunk iPXE on 2 different boards (and different SoC vendor) and confirmed the issue.

I'd like to see if there is anyone else uses iPXE on ARM platform, and if there exists some integration tests for iPXE that I can play with locally. By integration test I mean something like QEMU full ARM64 system emulation to load U-Boot/EDK2 and then load iPXE. Thanks.

**Comment by NiKiZe** [comment, marked as answer]
I think there is other issues/discussions about input on ARM.
If I remember correctly this is due to lack of timing events from the EFI firmware.

**Reply by mcb30** [reply]
This is correct.  Specifically, the U-Boot implementation of UEFI fails to implement timer interrupts, which means that everything freezes as soon as iPXE calls `cpu_nap()` since no interrupt will ever wake up the CPU.

As a workaround for this U-Boot bug, you can use `config/local/nap.h` to select `NAP_NULL`:
```
#undef NAP_EFIARM
#define NAP_NULL
```

This will cause iPXE to leave the CPU spinning instead of putting it to sleep, and will avoid the problems on U-Boot.

**Reply by frezbo** [reply]
Interesting, iPXE seems to work fine on Raspberry Pi's so there could be something different in the u-boot builds for each

**Reply by mcb30** [reply]
The Raspberry Pi is supported directly by EDK2 and so the UEFI implementation does not involve U-Boot in any way.

**Reply by RadxaYuntian** [reply]
Verified with our Radxa Zero 2 (based on Amlogic A311D). Thanks!

**Reply by benthetechguy** [reply]
Confirmed that this issue also takes place on ARMv7 with U-Boot. Should we make NAP_NULL default on arm64 and arm32 builds? What if this also takes place on x86_64 U-Boot? I haven't been able to test that yet.

**Reply by NiKiZe** [reply]
As stated above, it works as it should on all CPUs but is dependant on firmware implementation. As such it is not a solution to change the default based on CPU.

**Comment by frezbo** [comment]
Found some old issue in iPXE, I'm not sure if this is relevant:
- https://github.com/ipxe/ipxe/issues/165
- https://ipxe-devel.ipxe.narkive.com/wkDSSmnD/patch-1-1-efi-provide-possibility-to-disable-cpu-nap

**Reply by mcb30** [reply]
>     * [Arm64 build with serial console missing symbols. #165](https://github.com/ipxe/ipxe/issues/165)

This is not relevant.

>     * https://ipxe-devel.ipxe.narkive.com/wkDSSmnD/patch-1-1-efi-provide-possibility-to-disable-cpu-nap

This was cleaned up and became the `NAP_NULL` option that I described above.
